### DISCUSSION TITLE
How to implement session state to use with RLS

### DISCUSSION THREAD

**Discussion by mgm1313** [supabase/supabase#6301] [Questions]
I'm developing a multi-tenancy app, in which a single auth user can belong to multiple tenants. 
Inside the app, users can select from a list of tenants the tenant they want to view the data for. 

**Let's take any mail client as an example**. 

The database schema is set up as that all emails live inside one table: `email`, which besides columns for `sender`, `subject` and `content` each row is given a value in the `tenant_id` column. 
 
Say user Tim exists as one row in the `auth.users` table, but maintains two inboxes that of tenant `A` and tenant `B`, with corresponding ids. We store the tenant ids what Tim has access to as an array in the [app_metadata field](https://supabase.github.io/gotrue-js/interfaces/AdminUserAttributes.html#app_metadata).

Now we want to build a view in which Tim can view either the emails of tenant A or tenant B. I see a view different strategies for this:

**A. Filter client side**
We setup our RLS policies to only return `email` rows where the `tenant_id` value of that row is included in the `app_metadata` array. On the client, inside the Supabase query, we filter the rows further to only select those that we need (either tenant A or B). 

👍 Simple to implement, because we keep all state on the client. 
👍 Easy to later create views in which data of multiple inboxes can be combined (eg. show all mails of tenant A _and_ B).

👎 As the application grows, it's cumbersome to include the filter in every query. 
👎 Furthermore there is a high risk of forgetting to set the filter, in which case we could end up with mixed data of the two accounts, possibly without the user realizing. 

**B. Store in user_metadata** 
Another way would be to update the [users metadata](https://supabase.com/docs/reference/javascript/auth-update#update-password-for-authenticated-user) when they select a different tenant from the context menu. In our RLS policy we now first check if the passed `tenant_id ` inside the `user_metadata` is also part of the `app_metadata` array, so we prevent malicious use. Then we return only the `email` rows for which the tenant_id is equal to that inside the `user_metadata`. 

👍 We get rid of all the filter logic on the client, and rely on our RLS policies (which we already did for separating tenants which the user does not have access to). 
👍 Also works with realtime!

👎 It becomes harder to have aggregate views of all tenants the user has access to, and separate views on individual tenant level (eg. a massive inbox with all tenants emails vs. single tenant inboxes, within the same app).
 👎 When using multiple sessions (eg. devices) it gives a weird experience, since the `user_metadata` is not only stored in the JWT, but also in the database. So say a user is logged in on device 1 and device 2. On device 1 we switch from tenant A to tenant B, nothing happens at this point on device 2 since the JWT is still valid. But the moment the JWT of device 2 is refreshed, the context of that device now also changes to tenant B, without the user explicitly defining so. 

**C. Send as a header**
A third option I can think of is [setting a custom header](https://supabase.com/docs/reference/javascript/initializing#with-additional-parameters) when initializing the Supabase client, in which we include the current tenant_id the same way as in option B. When we change context inside the application (eg. change the inbox) we re-initialize the Supabase client with the new header. 

👍 We get both rid of the filter logic on the client, as wel as create a truly session specific solution (e.g. changing the header on device A won't influence device B). 

👎 Doesn't work with realtime, since web sockets [don't allow setting custom headers](https://github.com/supabase/realtime-js/issues/108#issuecomment-1000573730). 
👎 I'm not quite sure of unintended side effects when throwing away and recreating the Supabase instance in the application for stuff like `onAuthStateChange` listeners. 

Options B and C seem most promising, and have used to both so far. I really like the fact that we're able to keep on using realtime with option B, but it's a very bad UX that the tenant_id changes over all sessions. 

I'm curious as to how others have possibly approached this problem, or if there are some other options that I'm missing :)

**Comment by psteinroe** [comment]
Hey,

thanks for the write up!

We are currently looking to improve on the same area. A similar discussion is located here https://github.com/supabase/supabase/discussions/1148

As already mentioned there, maybe Supabase could allow us to set the `db-pre-request` (https://postgrest.org/en/stable/configuration.html#db-pre-request) option of PostgREST to set user metadata such as the tenant id when the request hits the database.

EDIT: It seems like there already is an issue for this https://github.com/supabase/supabase/issues/3233, although with another use-case.

**Comment by psteinroe** [comment]
I just received answer from the (super awesome) supabase support and they hinted that db-pre-request can already be set!

```sql
ALTER ROLE authenticator SET pgrst.db_pre_request to 'your_custom_function';
NOTIFY pgrst, 'reload config';
```

and to reset

```sql
ALTER ROLE authenticator RESET pgrst.db_pre_request;
NOTIFY pgrst, 'reload config';
```

more details [here](https://postgrest.org/en/stable/configuration.html#in-database-configuration)!

**Reply by mgm1313** [reply]
Cool, thanks for sharing this info! 

I'm doubting a bit where this approach with the db-pre-request would fit inside the already mentioned three approaches. It seems to have quite some overlap with approach B, but without the ability to be able to use Realtime (since that one doesn't use PostgREST). And I'm also not quite sure if it would help getting truly individual session state, instead of user wide state. 
Maybe you have some ideas about this?

**Reply by psteinroe** [reply]
You are right. Especially since its not working for Realtime it will only support the options described but not fix anything. At least, it will simplify the approach proposed in https://github.com/supabase/supabase/discussions/1148.

**Comment by reyhaan** [comment]
I am also in the same boat, I really want to use supabase as it will solve a TON of other problems for me to build my saas but it's missing the multi tenant use case.

I was playing with a design where I add a server in between supabase and clien, this server might take care of different sessions, I don't even know if it would work or not but I would really appreciate it if we can solve this multi tenant use case.

**Comment by mgm1313** [comment]
I think the solution for me would be to set per-session custom claims in the JWT. That would be like using approach B, but without the claims being persisted to the auth user. That way we achieve truly per-session state, while keeping the realtime option.
I opened a feature request in https://github.com/supabase/gotrue-js/issues/268

**Comment by dinvlad** [comment]
Any updates on this one? Firebase/GCIP Auth has multi-tenancy as a [first-class](https://firebase.google.com/docs/reference/admin/node/firebase-admin.auth.tenant) citizen, and with the same ultimate solution (I think) as proposed by @mgm1313 - meaning, `tenantId` is assigned as a temporary JWT claim upon each user sign-in or token refresh.
